proposals/0095-xof-hashing.md

```diff
@@ -0,0 +1,325 @@
+customize the output length and incorporate personalized domain separation 
+parameters. It operates by taking a message `N` and a customization string `S` 
+as input, enabling users to generate hash outputs of variable lengths `L` and 
+tailor the hashing process based on specific application requirements. 
+Extendable-output function's such as cSHAKE can be used for:
+
+- *Customized hashing:* cSHAKE's ability to produce variable-length hash 
+outputs makes it valuable for applications requiring flexible and tailored 
+hashing functions, such as in blockchain protocols, where different data 
+structures might require different hash lengths.
+- *Domain Separation:* It is beneficial in situations where secure and 
+domain-separated hashing is necessary, like in cryptographic protocols and 
+systems where unique hashing based on different contexts or domains is crucial.
+- *Protocols:* It is useful in protocols such as the 
+[STROBE protocol](https://strobe.sourceforge.io/papers/strobe-latest.pdf), 
+which is a versatile protocol 
+framework used to construct cryptographic primitives by composing different 
+operations in a sequence known as a protocol transcript. It allows for flexible 
+and efficient design of cryptographic protocols by assembling operations like 
+hashing, encryption, and authentication in a customizable manner. 
+
+Integrating cSHAKE would enhance Solana's cryptography toolkit, enabling 
+developers to create more secure and flexible applications on the platform. By 
+incorporating cSHAKE, Solana can leverage the full potential of the 
+[BulletProofs](https://github.com/dalek-cryptography/bulletproofs) 
+zero-knowledge proof library, enabling the efficient generation and 
+verification of non-interactive proofs. Applications involving 
+privacy-preserving transactions, such as confidential asset transfers, can 
+leverage Bulletproofs for efficient range proofs, while cSHAKE provides 
+customizable hashing for enhanced security.
+
+Overall, integrating cSHAKE and enabling to build the Bulletproofs 
+zero-knowledge proof library into Solana's infrastructure broadens the 
+platform's cryptographic capabilities, fostering enhanced privacy, security, 
+and flexibility for a wide array of decentralized applications and use cases.
+
+## Alternatives Considered
+
+Rewriting the BulletProof zero-knowledge library such that the merlin 
+transcripts are not based on any extendable output function. However, that 
+would change the security guarantees, and is most probably more complicated to 
+implement. 
+
+Another alternative is to implement the BulletProof zero-knowledge library as a 
+native program entirely, however, this would limit the use cases that can 
+additionally be enabled by supporting the customable extendable output functin 
+cSHAKE, and the merlin transcripts. Though, supporting a native zero-knowledge 
+proof library would likely be more efficient. 
```

Comment:
I will re-double this message. Adding new syscalls can hurt overall performance for all smart contracts. It would be useful to see an SBPF implementation as that may be sufficient, especially if only one or a handful of apps need this right now.

Reply:
I think there are much more benefits that other developers could have from existing cSHAKE syscalls. A cSHAKE implementation would enable to build customisable output length hash functions, domain separation for different protocol components and potential higher security due to customisation features. Overall it would make it a more versatile and potentially robust choice for some cryptographic operations within the Solana VM.